include/aws/common/ring_buffer.h

```diff
@@ -74,6 +65,30 @@ AWS_STATIC_IMPL bool aws_ring_buffer_check_atomic_ptr(
     return (atomic_ptr >= ring_buf->allocation && atomic_ptr <= ring_buf->allocation_end);
 }
 
+/**
+ * Checks whether the ring buffer is empty
+ */
+AWS_STATIC_IMPL bool aws_ring_buffer_is_empty(const struct aws_ring_buffer *ring_buf) {
+    uint8_t *head = (uint8_t *)aws_atomic_load_ptr(&ring_buf->head);
+    uint8_t *tail = (uint8_t *)aws_atomic_load_ptr(&ring_buf->tail);
+    return head == tail;
+}
+
+/**
+ * Evaluates the set of properties that define the shape of all valid aws_ring_buffer structures.
+ * It is also a cheap check, in the sense it run in constant time (i.e., no loops or recursion).
+ */
+AWS_STATIC_IMPL bool aws_ring_buffer_is_valid(const struct aws_ring_buffer *ring_buf) {
+    uint8_t *head = (uint8_t *)aws_atomic_load_ptr(&ring_buf->head);
+    uint8_t *tail = (uint8_t *)aws_atomic_load_ptr(&ring_buf->tail);
+    bool head_in_range = aws_ring_buffer_check_atomic_ptr(ring_buf, head);
+    bool tail_in_range = aws_ring_buffer_check_atomic_ptr(ring_buf, tail);
+    /* if head points-to the first element of the buffer then tail must too */
+    bool valid_head_tail = (head != ring_buf->allocation) || (tail == ring_buf->allocation);
+    return ring_buf && AWS_MEM_IS_READABLE(ring_buf->allocation, ring_buf->allocation_end - ring_buf->allocation) &&
```

Comment:
Can this subtraction underflow?

Reply:
I don't believe so since we could prove the invariant `allocation < allocation_size`. These variables are set once by a successful ring_buffer init() and then never modified. There is a precondition (that I think you added?) that the size of the requested size of the ring_buffer is non-zero and `allocation_end` is set to `allocation+size`.